ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -73,46 +114,72 @@ static double started_tracing_at_microseconds = 0;
+static VALUE tracing_start(VALUE _self, VALUE output_path);
+static VALUE tracing_stop(VALUE _self);
+static double timestamp_microseconds(void);
+static void set_native_thread_id(thread_local_state *);
+static void render_event(thread_local_state *, const char *event_name);
+static void on_thread_event(rb_event_flag_t event, const rb_internal_thread_event_data_t *_unused1, void *_unused2);
+static void on_gc_event(VALUE tpval, void *_unused1);
+static VALUE mark_sleeping(VALUE _self);
+
 void Init_gvl_tracing_native_extension(void) {
+#ifdef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET // 3.3+
+  thread_storage_key = rb_internal_thread_specific_key_create();
+#endif
+
   rb_global_variable(&gc_tracepoint);
 
   VALUE gvl_tracing_module = rb_define_module("GvlTracing");
 
+  rb_define_singleton_method(gvl_tracing_module, "_init_local_storage", tracing_init_local_storage, 1);
   rb_define_singleton_method(gvl_tracing_module, "_start", tracing_start, 1);
   rb_define_singleton_method(gvl_tracing_module, "_stop", tracing_stop, 0);
   rb_define_singleton_method(gvl_tracing_module, "mark_sleeping", mark_sleeping, 0);
 }
 
-static inline void initialize_thread_id(void) {
-  current_thread_seen = true;
-  current_thread_serial = RUBY_ATOMIC_FETCH_ADD(thread_serial, 1);
-  set_native_thread_id();
+static inline void initialize_thread_id(thread_local_state *state) {
+  state->current_thread_seen = true;
+  state->current_thread_serial = RUBY_ATOMIC_FETCH_ADD(thread_serial, 1);
+  set_native_thread_id(state);
 }
 
-static inline void render_thread_metadata(void) {
+static inline void render_thread_metadata(thread_local_state *state) {
   char native_thread_name_buffer[64] = "(unnamed)";
 
   #ifdef HAVE_PTHREAD_GETNAME_NP
     pthread_getname_np(pthread_self(), native_thread_name_buffer, sizeof(native_thread_name_buffer));
   #endif
 
+#ifdef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET
+  uint64_t thread_id = (uint64_t)state->thread;
+#else
+  uint64_t thread_id = state->thread_id;
+#endif
   fprintf(output_file,
     "  {\"ph\": \"M\", \"pid\": %"PRId64", \"tid\": %"PRIu64", \"name\": \"thread_name\", \"args\": {\"name\": \"%s\"}},\n",
     process_id, thread_id, native_thread_name_buffer);
 }
 
+static VALUE tracing_init_local_storage(UNUSED_ARG VALUE _self, VALUE thread) {
+  GT_LOCAL_STATE(thread, true);
+  return Qtrue;
+}
+
 static VALUE tracing_start(UNUSED_ARG VALUE _self, VALUE output_path) {
   Check_Type(output_path, T_STRING);
 
   if (output_file != NULL) rb_raise(rb_eRuntimeError, "Already started");
   output_file = fopen(StringValuePtr(output_path), "w");
   if (output_file == NULL) rb_syserr_fail(errno, "Failed to open GvlTracing output file");
 
+  thread_local_state *state = GT_CURRENT_THREAD_LOCAL_STATE();
   started_tracing_at_microseconds = timestamp_microseconds();
   process_id = getpid();
 
   fprintf(output_file, "[\n");
-  render_event("started_tracing");
+  render_event(state, "started_tracing");
```

Comment:
Since the only place `state->thread` is set is in `on_thread_event`, this event always gets written to the file with a `tid` of `0` in ruby 3.3.

```
{"ph": "M", "pid": 31042, "tid": 0, "name": "thread_name", "args": {"name": ""}},
{"ph": "E", "pid": 31042, "tid": 0, "ts": 2.000000},
{"ph": "B", "pid": 31042, "tid": 0, "ts": 2.000000, "name": "started_tracing"},
```

Reply:
We now also save the thread as soon as we create the thread local storage, since we have the thread handy. That removes all the tid=0 lines.